<!DOCTYPE html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" >
</head>
<html>
<body>
 <div id="wikicontent">
 <div class="vt" id="wikimaincol">
 <p><img src="./images/pagan_logo.png" width="400px"> </p><p>v. 0.33 </p><p><br> </p><p>This is a snapshot of the PAGAN documentation at <a href="http://code.google.com/p/pagan-msa/wiki/PAGAN">the program homepage</a> and may be outdated. Please use the online version if possible. <br> </p><h1><a name="Introduction"></a>Introduction<a href="#Introduction" class="section_anchor"></a></h1><p>PAGAN is  a general-purpose method for the alignment of sequence graphs. The main advantage of the graph representation of sequences is the graphs&#x27; ability to describe the uncertainty in the presence of characters at certain sequence positions. This is highly useful in phylogenetic progressive alignment and allows for capturing the different properties of insertions and deletions  when alignments are iterated for multiple sequences. However, graphs also allow describing the uncertainty in input sequences and modelling e.g. homoploymer errors in Roche 454 reads, or representing inferred ancestral sequences against which other sequences can then be aligned. </p><p>The development of PAGAN started in <a href="http://www.ebi.ac.uk/goldman/" rel="nofollow">Nick Goldman&#x27;s group</a> at the EMBL-European Bioinfomatics Institute, UK, and continues now in <a href="http://wiki.helsinki.fi/display/loytynojalab/Home" rel="nofollow">Ari LÃ¶ytynoja&#x27;s group</a> at the Institute of Biotechnology, University of Helsinki, Finland. PAGAN is still under development and will hopefully evolve to an easy-to-use, general-purpose method for phylogenetic sequence alignment. The graph representation has features that make it especially powerful for phylogenetic placement of sequences into existing alignments. As such a tool has direct applications in the analyses of increasingly abundant RNA-seq data and there is no satisfactory solution available, we have focused on this problem and implemented the necessary functionality first.  </p><p>This document explains how to install and start using the PAGAN software.  </p><p><br> </p><p><ul><li><a href="#Introduction">Introduction</a></li><li><a href="#Installation_of_PAGAN">Installation of PAGAN</a></li><ul><li><a href="#Installation_of_Boost_libraries">Installation of Boost libraries</a></li><li><a href="#Download_and_installation_of_PAGAN_using_git">Download and installation of PAGAN using git</a></li></ul><li><a href="#Using_PAGAN">Using PAGAN</a></li><li><a href="#Phylogenetic_multiple_alignment">Phylogenetic multiple alignment</a></li><li><a href="#Phylogenetic_placement_of_sequences_into_an_existing_alignment">Phylogenetic placement of sequences into an existing alignment</a></li><ul><li><a href="#Assignment_of_reads_to_specific_nodes">Assignment of reads to specific nodes</a></li><li><a href="#Use_of_Archaeopteryx-PAGAN-Edition">Use of Archaeopteryx-PAGAN-Edition</a></li><li><a href="#Examples_of_guided_placement">Examples of guided placement</a></li><ul><li><a href="#Example_of_guided_placement_of_protein_sequences">Example of guided placement of protein sequences</a></li><li><a href="#Example_of_guided_placement_of_NGS_reads">Example of guided placement of NGS reads</a></li></ul><li><a href="#Unguided_placement">Unguided placement</a></li><li><a href="#Pair-end_reads_and_454/_PacBio_data">Pair-end reads and 454/ PacBio data</a></li><li><a href="#Overlapping_pair-end_reads">Overlapping pair-end reads</a></li><li><a href="#Consensus_ancestor_reconstruction_and_phylogenetic_contig_assemb">Consensus ancestor reconstruction and phylogenetic contig assembly</a></li><li><a href="#Typical_analysis_of_NGS_data">Typical analysis of NGS data</a></li></ul><li><a href="#Pileup_alignment_without_a_guide_tree">Pileup alignment without a guide tree</a></li><ul><ul><li><a href="#Example_of_pileup_alignment_of_454_reads">Example of pileup alignment of 454 reads</a></li></ul></ul><li><a href="#Additional_program_options">Additional program options</a></li><ul><li><a href="#Generic_options">Generic options</a></li><li><a href="#Phylogenetic_placement_options">Phylogenetic placement options</a></li></ul><li><a href="#Inference_of_ancestral_sequences_for_existing_alignments">Inference of ancestral sequences for existing alignments</a></li><li><a href="#Configuration_files:_input,_output_and_format">Configuration files: input, output and format</a></li><li><a href="#Checking_for_PAGAN_updates">Checking for PAGAN updates</a></li><li><a href="#Appendix:_installing_boost_from_source_code">Appendix: installing boost from source code</a></li></ul> </p><p><br> </p><hr/><p><br> </p><h1><a name="Installation_of_PAGAN"></a>Installation of PAGAN<a href="#Installation_of_PAGAN" class="section_anchor"></a></h1><p>PAGAN is open-source software licensed under the GPL. The source code is provided at <a href="http://code.google.com/p/pagan-msa" rel="nofollow">http://code.google.com/p/pagan-msa</a>. PAGAN is developed and tested on a Linux system and we can currently only provide instructions for that platform, with a special focus on the popular Ubuntu/Debian-based distributions.  </p><p><br> </p><h2><a name="Installation_of_Boost_libraries"></a>Installation of Boost libraries<a href="#Installation_of_Boost_libraries" class="section_anchor"></a></h2><p>PAGAN requires <i>-dev versions</i> of two utility libraries from <a href="http://www.boost.org" rel="nofollow">http://www.boost.org</a> that may not be included in standard OS installations. These libraries have to be installed <strong>before</strong> compiling the PAGAN source code. On Ubuntu, they can be installed using commands: </p><pre class="prettyprint">apt-cache search libboost-program-options
apt-cache search libboost-regex</pre><p>(See which version number, ending with <tt>-dev</tt>, is provided and edit the command below.) </p><pre class="prettyprint">sudo apt-get install libboost-program-options1.40-dev
sudo apt-get install libboost-regex1.40-dev</pre><p>If your distribution does not provide Boost libraries (highly  unlikely) or you are not allowed to install software on your system, you can download and install the necessary library from the Boost project repositories directly into your PAGAN source code directory. See Appendix for details. </p><p><br> </p><h2><a name="Download_and_installation_of_PAGAN_using_git"></a>Download and installation of PAGAN using <tt>git</tt><a href="#Download_and_installation_of_PAGAN_using_git" class="section_anchor"></a></h2><p>The most recent version of the PAGAN source code is available from the git-repository and snapshots of this are downloadable as compressed tar-packages.  The <tt>git</tt> software can be found at <a href="http://git-scm.com" rel="nofollow">http://git-scm.com</a>. On Ubuntu, it can also be installed using command: </p><pre class="prettyprint">sudo apt-get install git-core</pre><p>Given that <tt>git</tt> is installed, the PAGAN source code can be downloaded and compiled using commands: </p><pre class="prettyprint">git clone https://code.google.com/p/pagan-msa/
cd pagan-msa/src
make -f Makefile.no_Qt
./pagan</pre><p>Without <tt>git</tt>, the latest (but possibly not up-to-date) tar-packaged source code can be downloaded and compiled using commands: </p><pre class="prettyprint">wget http://pagan-msa.googlecode.com/files/pagan.src.latest.tgz
tar xzf pagan.src.latest.tgz
cd pagan-msa/src
make -f Makefile.no_Qt
./pagan</pre><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Using_PAGAN"></a>Using PAGAN<a href="#Using_PAGAN" class="section_anchor"></a></h1><p>PAGAN is a command-line program. It can be used by (a) specifying a list of options (command-line arguments) when executing the program, or (b) creating a configuration file with the  options and specifying that when executing the program. The configuration file does not need to be created from scratch as PAGAN can output the options specified for an analysis in a file. This file can then be edited if necessary and specified as the configuration file for another analysis. Alternatively, the file can be considered as a record of a particular analysis with a full description of options and parameters used. </p><p>A list of the most important program options is outputted if no arguments are provided: </p><pre class="prettyprint">./pagan</pre><p>and a more complete list is given with the option <tt>--help</tt>: </p><pre class="prettyprint">./pagan --help</pre><p>In general, the option names start with <tt>--</tt> and the option name and value (if any) are separated by a space. The configuration file makes an exception and can be specified without the option name: </p><pre class="prettyprint">./pagan option_file</pre><p>Also this one can be given in the standard format and the following command is equivalent: </p><pre class="prettyprint">./pagan --config-file option_file</pre><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Phylogenetic_multiple_alignment"></a>Phylogenetic multiple alignment<a href="#Phylogenetic_multiple_alignment" class="section_anchor"></a></h1><p>PAGAN is based on a progressive algorithm that aligns sequences according to a guide tree. It (currently) cannot compute a tree by itself and requires the user to provide a <strong>rooted</strong> binary tree relating the sequences. The leaf names in the tree and the sequence names (until the first space) in the sequence file have to match exactly. Alignment is only performed for the parts of the guide phylogeny that have sequences associated; the unnecessary branches and sequences are pruned/dropped out.  </p><p>The minimal command to perform the alignment is: </p><pre class="prettyprint">./pagan --seqfile sequence_file --treefile tree_file </pre><p>The resulting alignment will be written in file <tt>outfile.fas</tt>. If you want to use another file name, you can specify that with option <tt>--outfile</tt>: </p><pre class="prettyprint">./pagan --seqfile sequence_file --treefile tree_file --outfile another_name</pre><p>PAGAN will automatically add suffix <tt>.fas</tt>.  </p><p>The sequence input file has to be in FASTA format and the guide tree in Newick tree format, with branch lengths as substitutions per site. The resulting alignment will be written in FASTA format (and in XML-based <a href="http://www.ebi.ac.uk/goldman-srv/hsaml" rel="nofollow">HSAML format</a> if option <tt>--xml</tt> is used). PAGAN supports the alignment of nucleotide, amino-acid and codon sequences although the last two are still experimental. </p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Phylogenetic_placement_of_sequences_into_an_existing_alignment"></a>Phylogenetic placement of sequences into an existing alignment<a href="#Phylogenetic_placement_of_sequences_into_an_existing_alignment" class="section_anchor"></a></h1><p>There are many methods to add sequences to existing alignments. However, most of these ignore the phylogenetic relationships among the data and simply place the new sequences in the bottom of the alignment. The idea of phylogenetic placement is to extend existing alignments by aligning the new sequences to their true evolutionary positions in the reference alignment. This ensures that the new sequences are compared to the closest possible reference sequence and the alignment created, including the gaps opened for insertions and deletions, is as accurate as possible. </p><p>PAGAN can reconstruct ancestral sequence history for a reference alignment related by a phylogeny. Sequences can then be added to this reference alignment by aligning the new sequences against terminal (extant reference sequences) or internal nodes (reconstructed ancestral sequences) and making room for the added sequences in their correct evolutionary positions among the reference sequences; the addition of new sequences does not affect the relative alignment of original sequences. </p><p>The current version of PAGAN is primarily aimed for guided placement of sequences, i.e. addition of sequences for which the (rough) phylogenetic location is known. The program includes support for unguided placement and can exhaustively search for the best nodes to place the new sequences. However, the latter is still experimental and, if target nodes for guided placement are not specified and exhaustive search is not selected, PAGAN will by default add the sequences at the root of the phylogeny.  </p><p>Originally the target application for the phylogenetic placement was in the analysis of short NGS reads but the same principles also work for full-length DNA sequences and for amino-acid sequences. For historical reasons, the new sequences added to the alignment are called &#x27;reads&#x27;.  </p><p>The minimal command to perform the sequence placement is: </p><pre class="prettyprint">./pagan --ref-seqfile ref_alignment_file --ref-treefile ref_tree_file --readsfile reads_file</pre><p>The reference alignment has to be in FASTA format and the reference tree in Newick or Newick eXtended (NHX) format; the reads (or sequences in general) that will be added in the reference alignment can be either in FASTA or FASTQ format. If the reference alignment consists of one sequence only, the reference guide tree is not required. Again, you can define your own output file with option <tt>--outfile</tt>.   </p><p><br> </p><h2><a name="Assignment_of_reads_to_specific_nodes"></a>Assignment of reads to specific nodes<a href="#Assignment_of_reads_to_specific_nodes" class="section_anchor"></a></h2><p>If the phylogenetic position of the sequences to be added is known (e.g. the species in question is an outgroup for the clade X), this information can be provided for the placement. Furthermore, one can name multiple alternative positions for the sequences (because the precise position is not known or the reference contains paralogous genes and the copies which the sequences come from are not known) and PAGAN will choose the one where the sequences matches the best. Sequences are assigned to a specific node or a set of nodes using the NHX tree format with an additional tag <tt>TID</tt>. The tag identifier can be any string, in the example below we use number <tt>001</tt>. One can use any number/combination of tags as long as each node and each sequence has at most one tag. </p><p>As an example, let&#x27;s assume  </p><p>a reference alignment: </p><pre class="prettyprint">&gt;A
AGCGATTG
&gt;B
AACGATCG
&gt;C
TGCGGTCC</pre><p>and a read that we want to add in FASTA: </p><pre class="prettyprint">&gt;D TID=001
AGCGATCG</pre><p>or in FASTQ format: </p><pre class="prettyprint">@read_D_0001@3@15@13524@18140#0/1 TID=001
AGCGATCG
+
CCCCCCCC</pre><p>The placement of the read in the reference alignment depends on the format of the reference tree:  </p><p>a phylogeny in plain Newick format adds the read at the root of the tree: </p><pre class="prettyprint">((A:0.1,B:0.1):0.05,C:0.15);</pre><p>a phylogeny in NHX format with one matching label adds the read to that node: </p><pre class="prettyprint">((A:0.1,B:0.1):0.05[&amp;&amp;NHX:TID=001],C:0.15);</pre><p>a phylogeny in NHX format with several matching labels finds the best node and adds the read to that: </p><pre class="prettyprint">((A:0.1,B:0.1):0.05[&amp;&amp;NHX:TID=001],C:0.15):0[&amp;&amp;NHX:TID=001];</pre><p>With ArchaeopteryxPE, the target nodes are shown with yellow dots and the three tree files look like this: </p><p><img src="./images/tree1.png" width="200px"> <img src="./images/tree2.png" width="200px"> <img src="./images/tree3.png" width="200px"> </p><p>In all cases, options <tt>--test-every-internal-node</tt> and <tt>--test-every-node</tt> override the placement information given in the guide tree and exhaustively search the best placement. </p><p><br> </p><h2><a name="Use_of_Archaeopteryx-PAGAN-Edition"></a>Use of Archaeopteryx-PAGAN-Edition<a href="#Use_of_Archaeopteryx-PAGAN-Edition" class="section_anchor"></a></h2><p>The NHX phylogeny format used for the tagging of target nodes for guided placement is unique for PAGAN. To facilitate the preparation and visualisation of the reference phylogenies, we have modified Archaeopteryx, the superb phylogeny visualisation program by Christian M Zmasek, to understand the PAGAN-specific tags and allow for their editing. The original Archaeopteryx program and documentation for its use can be found at <a href="http://phylosoft.org/archaeopteryx/" rel="nofollow">http://phylosoft.org/archaeopteryx/</a>. </p><p>The Archaeopteryx-PAGAN-Edition (here abbreviated as ArchaeopteryxPE) has one additional feature compared to the original program: the program can read, display, edit and write NHX files containing <tt>TID=&lt;string&gt;</tt> tags used for guided sequence placement with PAGAN. For any other purposes of tree display or editing you should use an up-to-date version of the original Archaeopteryx program. ArchaeopteryxPE is available as <a href="http://pagan-msa.googlecode.com/files/archaeopteryxPE.jar" rel="nofollow">Java jar-file</a> at the <a href="http://code.google.com/p/pagan-msa/downloads" rel="nofollow">download page</a> and can then be started with the command: </p><pre class="prettyprint">java -jar archaeopteryxPE.jar [reference.nhx]</pre><p>where the optional argument &#x27;reference.nhx&#x27; specifies a phylogeny file. (An example of a <a href="http://pagan-msa.googlecode.com/files/reference.nhx" rel="nofollow">valid tree file</a> is available at the <a href="http://code.google.com/p/pagan-msa/downloads" rel="nofollow">download page</a>.) This command assumes that the ArchaeopteryxPE program is located in the same directory; if not, type  </p><pre class="prettyprint">java -jar /path/to/dir/with/archaeopteryxPE.jar reference.nhx</pre><p>The source files modified to provide the necessary functionality are included in a <a href="http://pagan-msa.googlecode.com/files/archaeopteryxPE.modified_files.src.tgz" rel="nofollow">separate file</a>. This is provided to comply with the GPL and is not needed by a typical user. </p><p>The tree in the example file provided contains two types of NHX tags, <tt>D=Y</tt> and <tt>TID=0[12]</tt>. The red dots indicate duplication nodes (<tt>D=Y</tt>) and yellow dots the PAGAN target nodes (<tt>TID=01</tt> or <tt>TID=02</tt>). If a yellow node is clicked, a panel with the node data, including the PAGAN TID, is opened. To edit an existing TID tag or add a new tag at another node, the node has to be right-clicked and then &quot;Edit Node Data&quot; selected (or &quot;Edit Node Data&quot; selected at the &quot;Click on Node to:&quot; drop-down menu on the left and then the node clicked). This will open an edit panel with the PAGAN TID field in the bottom. The text-insert mode can be activated by clicking (repeatedly) the text field on the right side of the page icon. When saving the tree file, the NHX format can be selected from the file format drop-down menu. </p><p><br> </p><h2><a name="Examples_of_guided_placement"></a>Examples of guided placement<a href="#Examples_of_guided_placement" class="section_anchor"></a></h2><p>The PAGAN download files contain example datasets for guided placement of protein sequences (full length and fragments) and NGS reads. The datasets can found in directories <a href="http://code.google.com/p/pagan-msa/source/browse/#git%2Fexamples%2Fprotein_placement" rel="nofollow">`examples/protein_placement`</a> and <a href="http://code.google.com/p/pagan-msa/source/browse/#git%2Fexamples%2Fngs_placement" rel="nofollow">`examples/ngs_placement`</a>. </p><p>The idea of guided placement is to extend existing alignments with new sequences for which the (rough) phylogenetic location is known. The situation may be complicated by paralogous genes and uncertainty of the right target for the new sequences. For such cases, PAGAN allows indicating several potential targets and then chooses the best location by aligning the query sequences to each specified target. </p><p>The examples are based on simulated data. In these examples, we have reference alignments that include gene sequences from several mammals; the hypothetical gene has duplicated first in the ancestor of primates and rodents and then again in the ancestor of rodents. Our new sequences come from a primitive primate, believed to belong phylogenetically between tarsier and lemur, and a rodent, belonging between guinea pig and ground squirrel. Because of the gene duplications, the primate and rodent sequences have two and three potential target locations, respectively, in the phylogeny. These are indicated in the reference phylogenies using special TID tags. The phylogenies and the target locations for guided placement can be viewed and edited with the <a href="http://pagan-msa.googlecode.com/files/archaeopteryxPE.jar" rel="nofollow">ArchaeopteryxPE software</a> available on the <a href="http://code.google.com/p/pagan-msa/downloads" rel="nofollow">PAGAN download site</a>. ArchaeopteryxPE indicates the duplication nodes with red dots and the TID target nodes with yellow dots and the guide phylogeny looks like this: </p><p><img src="./images/guide_tree.png" width="600px"> </p><p>When sequences are placed with PAGAN, it first tries placing each query sequence to every matching target node and chooses the best node. If several target nodes score equally well, the sequence is assigned to each one of them. PAGAN then starts aligning the sequences to targets, adding several query sequences to the same target using a progressive approach. The complete new subtree is then inserted back to the reference alignment, adding space for the new insertions if necessary. Importantly, the relative alignment of existing reference sequences is not changed.  </p><p>The examples demonstrate that the selection of the target node for each query sequence is not straightforward. Long sequences are placed correctly but shorter ones may be placed to a wrong node. Errors are also more likely between closely-related paralogs than between distant ones. </p><p><br> </p><h3><a name="Example_of_guided_placement_of_protein_sequences"></a>Example of guided placement of protein sequences<a href="#Example_of_guided_placement_of_protein_sequences" class="section_anchor"></a></h3><p>The following files are included in directory <a href="http://code.google.com/p/pagan-msa/source/browse/#git%2Fexamples%2Fprotein_placement" rel="nofollow">`examples/protein_placement`</a>:  </p><ul><li><tt>reference_aa.fas</tt>   : simulated reference alignment. </li><li><tt>reference_tree.nhx</tt> : reference phylogeny with some nodes tagged for placement. </li><li><tt>input_aa_full.fas</tt>  : simulated amino-acid sequences. </li><li><tt>input_aa_frags.fas</tt> : a subset of the one above, broken into fragments. </li></ul><p>The data can be analysed using the following commands: </p><pre class="prettyprint">pagan --ref-seqfile reference_aa.fas --ref-treefile reference_tree.nhx \
 --readsfile input_aa_full.fas --outfile aa_full_alignment

pagan --ref-seqfile reference_aa.fas --ref-treefile reference_tree.nhx \
 --readsfile input_aa_frags.fas --outfile aa_frags_alignment</pre><p>The resulting alignments will be written to files <tt>aa_[full|frags]_alignment.fas</tt>.  </p><p>The reference alignment used for the analysis, the true simulated alignment and the resulting PAGAN alignments look like this (click to see larger images):    </p><p><table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/reference_aa.png" rel="nofollow"><img src="./images/reference_aa_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Reference alignment </td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/aa_true_alignment.png" rel="nofollow"><img src="./images/aa_true_alignment_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> True simulated alignment </td></tr> </table></p><p><table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/aa_full_alignment.png" rel="nofollow"><img src="./images/aa_full_alignment_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Placement of full sequences </td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/aa_frags_alignment.png" rel="nofollow"><img src="./images/aa_frags_alignment_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Placement of sequence fragments </td></tr> </table></p><p>The different background colours for the placed sequences indicate their different origin.    <br> </p><h3><a name="Example_of_guided_placement_of_NGS_reads"></a>Example of guided placement of NGS reads<a href="#Example_of_guided_placement_of_NGS_reads" class="section_anchor"></a></h3><p>The following files are included in directory <a href="http://code.google.com/p/pagan-msa/source/browse/#git%2Fexamples%2Fngs_placement" rel="nofollow">`examples/ngs_placement`</a>: </p><ul><li><tt>reference_codon.fas</tt> : simulated reference alignment. </li><li><tt>reference_tree.nhx</tt>  : reference phylogeny with some nodes tagged for placement. </li><li><tt>input_ngs.fastq</tt>     : simulated Illumina reads. </li><li><tt>input_ngs_primates.fastq</tt> : a subset of the one above. </li></ul><p>The data can be analysed using the following command: </p><pre class="prettyprint">pagan --ref-seqfile reference_codon.fas --ref-treefile reference_tree.nhx \
 --readsfile input_ngs.fastq --outfile read_alignment</pre><p>The resulting alignments will be written to file read_alignment.fas. </p><p>If we add the option <tt>--config-log-file</tt>: </p><pre class="prettyprint">pagan --ref-seqfile reference_codon.fas --ref-treefile reference_tree.nhx \
--readsfile input_ngs.fastq --outfile read_alignment --config-log-file simple.cfg</pre><p>the arguments used for the analysis will be written to the file <tt>simple.cfg</tt>. This file works as a log file of the specific settings used for the analysis and allows repeating the same analysis with the command: </p><p><tt>pagan simple.cfg</tt>  (or <tt>pagan --config-file simple.cfg</tt>) </p><p>The use of config files simplifies complex commands. For example, the following:   </p><pre class="prettyprint">pagan --ref-seqfile reference_codon.fas --ref-treefile reference_tree.nhx \
--readsfile input_ngs.fastq --trim-read-ends --build-contigs --use-consensus \
--consensus-minimum 3 --show-contig-ancestor --outfile read_alignment \
--config-log-file contigs.cfg</pre><p>specifies config file <tt>contigs.cfg</tt>. As the arguments given on the command line override those given in a config file, the following command repeats a similar analysis for another data set. </p><pre class="prettyprint">pagan contigs.cfg --readsfile input_ngs_primates.fastq --outfile primate_alignment</pre><p>The reference alignment used for the analysis, the true simulated alignment and the resulting PAGAN alignments look like this:    </p><p><table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/reference_codon.png" rel="nofollow"><img src="./images/reference_codon_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Reference alignment </td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/true_alignment.png" rel="nofollow"><img src="./images/true_alignment_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> True simulated alignment </td></tr> </table></p><p><table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/read_alignment.png" rel="nofollow"><img src="./images/read_alignment_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Placement of all reads </td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/primate_alignment.png" rel="nofollow"><img src="./images/primate_alignment_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Placement of primate reads </td></tr> </table></p><p>The different background colours for the placed sequences indicate their different origin. </p><p>You may check the content of files <tt>read_alignment_contigs.fas</tt> and  <tt>primate_alignment_contigs.fas</tt>:   <table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/read_alignment_contigs.png" rel="nofollow"><img src="./images/read_alignment_contigs_s.png" /></a></td><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/primate_alignment_contigs.png" rel="nofollow"><img src="./images/primate_alignment_contigs_s.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Contigs from all reads </td><td style="border: 1px solid #ccc; padding: 5px;"> Contigs from primate reads </td></tr> </table></p><p>The quality of contigs is expected to improve with a higher sequencing coverage.  </p><p>See <tt>pagan --help</tt> and the program documentation for additional options. </p><p><br> </p><h2><a name="Unguided_placement"></a>Unguided placement<a href="#Unguided_placement" class="section_anchor"></a></h2><p>If no target nodes (TID tags) are specified in the reference phylogeny, PAGAN will by default add the sequences in the bottom of the alignment by aligning them against the root node. This can be overridden with options <tt>--test-every-internal-node</tt> and <tt>--test-every-node</tt> that exhaustively search through either internal nodes or all nodes (including leaf nodes) and add the sequence at the node where it matches the best.  </p><p>Exhaustive search using PAGAN&#x27;s exact alignment can be very slow. To speed up the search of the optimal placement node, PAGAN can use Guy Slater&#x27;s <a href="http://www.ebi.ac.uk/~guy/exonerate" rel="nofollow">Exonerate</a>; Exonerate is not provided by the PAGAN package and has to be installed separately and be on the system&#x27;s execution path ($PATH). The default combination of settings is chosen with option <tt>--fast-placement</tt> and typically one would also use option <tt>--test-every-internal-node</tt> or <tt>--test-every-node</tt>, or have a guide tree with tagged nodes for placement. With this option, PAGAN runs Exonerate local alignment against the chosen nodes (all/all internal/all tagged) and keep the best of them; it would then run Exonerate gapped alignment against those nodes and choose the best for the placement.   With other options, one can choose to perform just a local or gapped alignment with Exonerate, and change the number of nodes kept after each round. The number of nodes kept is based either on the score relative to the best score, or a fixed number of best-ranking nodes. If more than one node is returned by Exonerate, PAGAN will perform alignments against those and choose the best scoring one for the placement. With option <tt>--use-exonerate-anchors</tt>, Exonerate hits are used to reduce the search space in the final sequence placement; this option is automatically chosen with <tt>--fast-placement</tt>. </p><p><br> </p><h2><a name="Pair-end_reads_and_454/_PacBio_data"></a>Pair-end reads and 454/PacBio data<a href="#Pair-end_reads_and_454/_PacBio_data" class="section_anchor"></a></h2><p>For the alignment of NGS reads, additional options <tt>--pair-end</tt>, <tt>--454</tt> and <tt>--pacbio</tt> are useful. The first one merges paired reads into one (separated by a spacer) before the alignment and the second and third model the ambiguous length of mononucleotide runs and the high frequency of insertion-deletions in data coming from Roche 454 and PacBio SMRT platforms. The pairing of pair-end reads assumes that the reads have identical names (until the first space) with the exception that the left read ends with <tt>/1</tt> and the right read with <tt>/2</tt>. In the resulting alignment, the paired sequence will have suffix <tt>/p12</tt>. </p><p><br> </p><h2><a name="Overlapping_pair-end_reads"></a>Overlapping pair-end reads<a href="#Overlapping_pair-end_reads" class="section_anchor"></a></h2><p>The length and overall quality of NGS data can be improved by using such a short fragment length that the reads starting from each end of the fragment overlap in the middle. PAGAN allows merging such reads and handling them as one sequence. Merging is done using option <tt>--overlap-pair-end</tt>, the reads successfully merged having suffix <tt>/m12</tt> in the resulting alignment.  </p><p>The merging requires significant overlap between the two reads in their pairwise alignment (performed without masking). Options <tt>--overlap-minimum</tt> and <tt>--overlap-identity</tt> can be used to change the minimum length and base identity of this overlap. Shorter overlaps that show perfect base identity are also accepted; the minimum length of this can be changed with option <tt>--overlap-identical-minimum</tt>. The merged reads can be outputted in FASTQ format using option <tt>--overlap-merge-file</tt>. PAGAN can also be used just to merge the overlapping reads: </p><pre class="prettyprint">./pagan --overlap-merge-only --readsfile reads_file --overlap-merge-file merge_file</pre><p><br> </p><h2><a name="Consensus_ancestor_reconstruction_and_phylogenetic_contig_assemb"></a>Consensus ancestor reconstruction and phylogenetic contig assembly<a href="#Consensus_ancestor_reconstruction_and_phylogenetic_contig_assemb" class="section_anchor"></a></h2><p>PAGAN&#x27;s graph reconstruction is based on maximum parsimony. In NGS read placement, the overlapping regions of reads placed to a specific target node are expected to be identical and any differences observed are (with the exception heterozygosity that we ignore) some sort of errors. For the placement of DNA sequences, PAGAN offers an alternative approach to reconstruct ancestor graphs based on majority consensus. This is selected with option <tt>--use-consensus</tt>. </p><p>Using a similar majority consensus approach, PAGAN can combine reads placed to a specific target into longer contigs. This is done with option <tt>--build-contigs</tt>; the minimum number of reads supporting a site in the contig can be changed with option <tt>--consensus-minimum</tt>. The contigs, with the reads supporting them, are written to a file named <tt>*_contigs.fas</tt>. </p><p>By default, the disconnected contigs are bridged with <tt>n</tt>&#x27;s, estimating the size of the contig gap from the outgroup ancestral sequence. With option <tt>--show-contig-ancestor</tt> they can be bridged using the actual ancestral sequences, the contig gaps indicated with lower case characters. </p><p><br> </p><h2><a name="Typical_analysis_of_NGS_data"></a>Typical analysis of NGS data<a href="#Typical_analysis_of_NGS_data" class="section_anchor"></a></h2><p>A typical command to perform the placement of NGS reads could be: </p><pre class="prettyprint">./pagan --ref-seqfile ref_alignment_file --ref-treefile ref_tree_file --readsfile reads_file \
    [--trim-read-ends] [--discard-overlapping-identical-reads] [--rank-reads-for-nodes] \
    [--pair-end] [--overlap-pair-end] [--454] </pre><p>where the optional (and some mutually exclusive) options are in square brackets.  </p><p>Depending on the output, the thresholds for masking and trimming can be adjusted using the relevant options explained below. The effect of trimming is obvious from the output; the bases written in lower case in the output were masked and considered as N&#x27;s during the alignment. A sequence of N&#x27;s matches any sequence and gives rubbish alignments; if that happens, you may need to raise the masking threshold or lower the trimming thresholds. </p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Pileup_alignment_without_a_guide_tree"></a>Pileup alignment without a guide tree<a href="#Pileup_alignment_without_a_guide_tree" class="section_anchor"></a></h1><p>PAGAN can make &quot;pileup&quot; alignments by adding the sequences in the order their appear in the input file. This is not recommended for distantly-related sequences but can be useful for highly similar sequences that require alignment. This may be relevant e.g. in the analysis of overlapping noisy reads from the same locus; if the reads have been  generated on Roche 454 or Ion Torrent platform, PAGAN&#x27;s ability to model the homopolymer errors is especially useful. </p><p>Like in any alignment also in a pileup alignment the consenus sequence (the alignment of sequences included so far) and the next sequence to be added should overlap. PAGAN can include sequences that do not overlap but the region in between has to be bridged by other sequences and thus the order of adding the sequences can be important. If the sequences are from the same species, the option to reconstruct the consensus sequence can also be useful. </p><p>The pileup alignment is done by selecting option <tt>--reads-pileup</tt> and defining the input file with <tt>--readsfile reads_file</tt>.  Typically one would also add option <tt>--use-consensus</tt> to specify that the character states are based on consensus, not the last-added and possibly incorrect sequence (e.g. due to contamination of the data), and to output the final consensus sequence. If the sequences are from Roche 454 or Pacific Biosciences platforms, one may also add option <tt>--454</tt> or <tt>--pacbio</tt> to tune the parameters specifically for those instruments; many other options are naturally compatible with the pileup alignment, too. </p><p>The many options needed for this can be defined in a config file. Assuming that the file <tt>454.cfg</tt> contains the following: </p><pre class="prettyprint">454 = 1
reads-pileup = 1
use-consensus = 1
build-contigs = 1
no-terminal-edges = 1
silent = 1</pre><p>the pile-up alignment for a set of 454 data can then be generated with the command: </p><pre class="prettyprint">./pagan 454.cfg --readsfile reads_file --outfile output_file</pre><p><br> </p><h3><a name="Example_of_pileup_alignment_of_454_reads"></a>Example of pileup alignment of 454 reads<a href="#Example_of_pileup_alignment_of_454_reads" class="section_anchor"></a></h3><p>The PAGAN download files contain an example dataset for pileup alignment of overlapping 454 reads. The dataset can found in directory <a href="http://code.google.com/p/pagan-msa/source/browse/#git%2Fexamples%2F454_pileup" rel="nofollow">`examples/454_pileup`</a>. The following file is included: </p><ul><li>454_reads.fas : simulated overlapping 454 reads </li></ul><p>The data can be analysed using the following command: </p><pre class="prettyprint">pagan --reads-pileup --use-consensus --454 --readsfile 454_reads.fas --outfile 454_reads_pagan</pre><p>The resulting alignment will be written to file 454_reads_pagan.fas. </p><p>Using the config file from above, a similar analysis could be performed with the command: </p><pre class="prettyprint">pagan 454.cfg --readsfile 454_reads.fas --outfile 454_reads_pagan</pre><p>The modelling of homopolymer error has an impact even in the alignment of relatively clean data:  </p><p><table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/454_reads_pagan.png" rel="nofollow"><img src="./images/454_reads_pagan_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> PAGAN alignment </td></tr> </table></p><p><table class="wikitable"><tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/454_reads_mafft.png" rel="nofollow"><img src="./images/454_reads_mafft_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/454_reads_muscle.png" rel="nofollow"><img src="./images/454_reads_muscle_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"><a href="./images/454_reads_clustalw.png" rel="nofollow"><img src="./images/454_reads_clustalw_m.png" /></a></td></tr> <tr><td style="border: 1px solid #ccc; padding: 5px;"> Alignments with other popular methods </td></tr> </table></p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Additional_program_options"></a>Additional program options<a href="#Additional_program_options" class="section_anchor"></a></h1><p>Some of the options printed by <tt>./pagan --help</tt> relate to unfinished features and may not function properly. Only the main options are documented here. </p><p><br> </p><h2><a name="Generic_options"></a>Generic options<a href="#Generic_options" class="section_anchor"></a></h2><ul><li>Option <tt>--silent</tt> minimises the output (doesn&#x27;t quite make it silent, though). </li></ul><ul><li>Options <tt>--indel-rate</tt> can be used adjust the insertion and deletion rates. Although it would be possible to consider the two processes separately, this has not been implemented yet. </li></ul><ul><li>Options <tt>--gap-extension</tt> and <tt>--end-gap-extension</tt> define the gap extension probability for regular and terminal gaps. For meaningful results, the latter should be greater (and, for pair-end data, equal to <tt>--pair-read-gap-extension</tt>). </li></ul><ul><li>Options <tt>--dna-kappa</tt> and <tt>--dna-rho</tt> affect the DNA substitution scoring matrix; base frequencies are estimated from the data. </li></ul><ul><li>Options <tt>--codons</tt> translates DNA sequences to codons and aligns them using the codon substitution model. (experimental) </li></ul><ul><li>Options <tt>--scale-branches</tt>, <tt>--truncate-branches</tt> and <tt>--fixed-branches</tt> override the branch lengths defined in the guide tree. By default, long branches are truncated to make the scoring matrix more informative; this can be prohibited with <tt>--real-branches</tt>. </li></ul><ul><li>Option <tt>--output-ancestors</tt> writes the parsimony-reconstructed ancestral sequences for the internal nodes of the tree. The tree indicating the nodes is written in <tt>outfile.anctree</tt>. </li></ul><ul><li>Option <tt>--config-file file_name</tt> specifies a config file. If an option is specified both in a config file and as a command-line argument, the latter one overrides the former. </li></ul><ul><li>Option <tt>--config-log-file file_name</tt> specifies a log file where (non-default) options used for the analysis are written.  The format is compatible with the option input and the file can be used a config file. </li></ul><p>There are many parameters related to &quot;insertion calling&quot;, the type and amount of phylogenetic information required to consider insertion-deletion as an insertion and thus prevent the later matching of those sites. These parameters are still experimental (although some of them are used and affect the resulting alignment) and will be described in detail later. </p><p><br> </p><h2><a name="Phylogenetic_placement_options"></a>Phylogenetic placement options<a href="#Phylogenetic_placement_options" class="section_anchor"></a></h2><p>Some options are only relevant for the placement of sequences into existing alignment and many of those require FASTQ or NHX formatted input. </p><ul><li>Option <tt>--qscore-minimum</tt> sets the Q-score threshold for sites to be masked (replaced with N&#x27;s for alignment, shown in lower case in the output); <tt>--allow-skip-low-qscore</tt> further allows skipping those sites (usefulness of this is uncertain). Masking is done by default (see <tt>./pagan --help</tt> for the default threshold) and can be disabled with option <tt>--no-fastq</tt> that prohibits all Q-score-based pre-processing done either by default or chosen by the user (e.g. trimming and 454-specific modelling). </li></ul><ul><li>Option <tt>--trim-read-ends</tt> enables the trimming of FASTQ reads;  <tt>--trim-mean-qscore</tt>, <tt>--trim-window-width</tt> and  <tt>--minimum-trimmed-length</tt> define the minimum Q-score and width for the sliding window and the minimum length for the trimmed read. Trimming progresses inwards from each end until the mean Q-score for the window exceeds the score threshold; if the read is shortened below the length threshold, it is discarded. The length of removed (trimmed) fragments is indicated in the sequence name field in the format <tt>P1ST\$l1:P1ET\$l2</tt> where <tt>\$l1</tt> and <tt>\$l2</tt> refer to the start and end of the read. If the sequences consists of a read pair, the trimming of the right-hand side read is similarly indicated using the tag <tt>P2</tt>. </li></ul><ul><li>Option <tt>--rank-reads-for-nodes</tt> ranks the sequences assigned to a node and aligns them in that order. The ranking is based on their score in the placement alignment used to decide between the alternative nodes. If sequences are assigned to one node each, one round of alignments against each node is performed to define the ranking before the final multiple alignment. </li></ul><ul><li>Reads fully embedded in other reads can be discarded: <tt>--discard-overlapping-reads</tt> identifies reads that overlap in the placement alignment and removes ones that are fully embedded in another one; <tt>--discard-overlapping-identical-reads</tt> extends this by checking that the embedded read is identical (on base level) to the longer read before discarding it; <tt>--discard-pairwise-overlapping-reads</tt> makes all the pairwise alignments to identify embedded reads. </li></ul><ul><li>Option <tt>--reads-distance</tt> sets the expected distance between the read and the pseudo-parent node (against which the read is aligned) and thus affects the substitution scoring used in the alignment. Having the distance very short (default), the alignment is stringent and expects high similarity. </li></ul><ul><li>Reads with too few sites aligned against sites of reference sequences are discarded. (The stringency of the alignment is set using the option above.) Options <tt>--min-reads-overlap</tt> and  <tt>--min-reads-identity</tt> set the required overlap and base identity for accepting the read. </li></ul><p>The rest of the options are either not important for basic use or self-explanatory (or both). </p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Inference_of_ancestral_sequences_for_existing_alignments"></a>Inference of ancestral sequences for existing alignments<a href="#Inference_of_ancestral_sequences_for_existing_alignments" class="section_anchor"></a></h1><p>The features required for phylogenetic placement of reads can be used without any reads, too. Command: </p><pre class="prettyprint">./pagan --ref-seqfile alignment_file --ref-treefile tree_file --xml</pre><p>writes the aligned sequences from <tt>alignment_file</tt> to files <tt>outfile.fas</tt> and <tt>outfile.xml</tt>, thus providing a tool to convert FASTA files to HSAML format. This maybe more interesting with option <tt>--output-ancestors</tt>. Command: </p><pre class="prettyprint">./pagan --ref-seqfile alignment_file --ref-treefile tree_file --output-ancestors</pre><p>includes the parsimony-reconstructed internal nodes in the FASTA output, providing a efficient tool to infer gap structure for ancestral sequences based on existing alignments. The tree indicating the ancestral nodes is written in <tt>outfile.anctree</tt>. </p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Configuration_files:_input,_output_and_format"></a>Configuration files: input, output and format<a href="#Configuration_files:_input,_output_and_format" class="section_anchor"></a></h1><p>Configuration files contain option names and values separated by <tt>=</tt> sign, one option per row. Rows starting with a hash sign <tt>#</tt> are comments and ignored. Thus, if the content of file <tt>config.cfg</tt> is: </p><pre class="prettyprint"># this is an uninformative comment
ref-seqfile = reference_alignment.fas
ref-treefile = reference_tree.nhx
readsfile = illumina_reads.fastq
outfile = read_alignment
trim-read-ends = 1</pre><p>the command: </p><pre class="prettyprint">./pagan config.cfg</pre><p>(or <tt>./pagan --config-file config.cfg</tt>) </p><p>is equivalent to: </p><pre class="prettyprint">./pagan --ref-seqfile reference_alignment.fas --ref-treefile reference_tree.nhx \
  --readsfile illumina_reads.fastq --outfile read_alignment --trim-read-ends</pre><p>By adding the option <tt>--config-log-file config.cfg</tt> in the command above, PAGAN creates a config file that is equivalent with the one above (with some more comments). Config files can, of course, be written or extended manually using the same format. One should note, however, that also boolean options need a value assigned, such as <tt>trim-read-ends = 1</tt> in the example above. If a boolean option is not wanted, it should not specified in the config file (or it should be commented out with a hash sign) as setting an option e.g. <tt>0</tt> or <tt>false</tt> does not disable it. </p><p>Options in a config file are overridden by re-defining them on command line. Thus, the command: </p><pre class="prettyprint">./pagan config.cfg --outfile another_name</pre><p>is the same as the one above except that the results will be placed to a file with another name. </p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Checking_for_PAGAN_updates"></a>Checking for PAGAN updates<a href="#Checking_for_PAGAN_updates" class="section_anchor"></a></h1><p>PAGAN runs locally and does not contact or report to any external server. The only exception is the command <tt>./pagan --version</tt> that checks if a newer version of the program is available for download. The actual upgrade has to be performed manually. </p><p><br> </p><p><a href="#Introduction">back to top</a> </p><hr/><p><br> </p><h1><a name="Appendix:_installing_boost_from_source_code"></a>Appendix: installing <tt>boost</tt> from source code<a href="#Appendix:_installing_boost_from_source_code" class="section_anchor"></a></h1><p>If pre-packaged files are not available for your platform or you are not entitled to install packages and cannot persuade your system administrator to do that, the Boost libraries necessary for the compilation of the PAGAN source code can be installed using following commands: </p><pre class="prettyprint"># make a temporary directory
mkdir ~/tmp_boost
cd ~/tmp_boost

# get the source code
wget http://sourceforge.net/projects/boost/files/boost/1.44.0/boost_1_44_0.zip/download
unzip boost_1_44_0.zip
cd boost_1_44_0

# fix the permission
chmod +x tools/jam/src/build.sh

# compile and install
sh ./bootstrap.sh --prefix=$PATH_TO_PAGAN_DIR/boost --with-libraries=program_options,regex
./bjam install

# check that it&#x27;s there and set to library path
ls $PATH_TO_PAGAN_DIR/boost/lib
export LD_LIBRARY_PATH=$PATH_TO_PAGAN_DIR/boost/lib:$LD_LIBRARY_PATH

# clean up
cd ../..
rm -r ./tmp_boost</pre><p>Note that you need to replace <tt>$PATH_TO_PAGAN_DIR</tt> with the appropriate file path. You also need to copy the line <tt>export LD_LIBRARY_PATH=...</tt> in your <tt>~/.bashrc</tt> or similar such that it will automatically be set for future sessions. </p>
 </div>
 </div>
 </td><tr>
</table>
 </div>

 </body>
</html>


